docs/conf.py

```diff
@@ -361,3 +373,190 @@ def _skip(self, word):
 # If true, do not generate a @detailmenu in the "Top" node's menu.
 #
 # texinfo_no_detailmenu = False
+
+
+LOG = logging.getLogger(__name__)
+
+
+class DDTraceReleaseNotesDirective(rst.Directive):
+    r"""
+    Directive class to handle ``.. ddtrace-release-notes::`` directive.
+
+    This directive is used to generate up to date release notes from
+    Reno notes in this repo.
+
+    This directive will dynamically search for all release branches that
+    match ``^[\d]+\.[\d+]``, and generate the notes for each of those
+    releases.
+
+    When generating notes for a given branch we will also search for the
+    best "earliest version" to use for that branch. For example, on a new
+    release branch with only prereleases we will resolve to the rc1 version
+    for that release. If there are any non-prereleases for that branch we will
+    resolve to the first non-rc release.
+    """
+
+    has_content = True
+
+    def __init__(self, *args, **kwargs):
+        super(DDTraceReleaseNotesDirective, self).__init__(*args, **kwargs)
+
+        self._repo = dulwich.repo.Repo.discover(".")
+        self._release_branch_pattern = re.compile(r"^[\d]+\.[\d]+")
+
+    @property
+    def _release_branches(self):
+        # type: () -> list[tuple[Version, str]]
+        r"""
+        Helper to get a list of release branches for this repo.
+
+        A release branch exists under refs/remotes/origin/ and matches the pattern::
+
+            r"^[\d]+\.[\d]+"
+
+        The results are a list of parsed Version's from the branch name (to make sorting/
+        comparing easier), along with the original ref name.
+        """
+        versions = []  # type: list[tuple[Version, str]]
+        for ref in self._repo.get_refs():
+            # We get the ref as bytes from dulwich, convert to str
+            ref = ref.decode()
+
+            # Ignore any refs that aren't for origin/
+            if not ref.startswith("refs/remotes/origin/"):
+                continue
+
+            # Get just the branch name omitting origin/
+            # and make sure it matches our pattern
+            version_str = ref[20:]
```

Comment:
20 is a magic number here, maybe do a `len("...")`?

Reply:
the alternative is something like:

```python
_, _, version_str = ref.partition("refs/remotes/origin/")
```

maybe that way is more clear?

Reply:
```suggestion
            _, _, version_str = ref.partition("refs/remotes/origin/")
```